elbepack/xmlpreprocess.py

```diff
@@ -143,6 +143,33 @@ def preprocess_initvm_ports(xml):
             forward.getparent().remove(forward)
 
 
+def preprocess_pkg_pinning(xml):
+    """Do search and replace on pkg attributes, replacing 'pin' with 'release-name'."""
+
+    errors = False
+    for pkg in xml.iterfind('//target/pkg-list/pkg'):
+        # Check that max one attribute of 'version', 'origin', 'release-* / pin' is set
+        if int('version' in pkg.attrib.keys()) \
+            + int('origin' in pkg.attrib.keys()) \
+            + max(len([a for a in pkg.attrib.keys() if a.startswith('release-')]) + int('pin' in pkg.attrib.keys()), 1) \
+             > 1:
+            logging.error(
+                'Invalid pkg pinning attribute combination for "%s".' \
+                ' Use only either of "version" OR "origin" OR "release-* | pin"',
+                pkg.text)
+            errors = True
+
+        if 'pin' in pkg.attrib.keys():
+            print('[WARN] attribute pin= for element <pkg> is deprecated. '
+                'Use release-name= instead.')
```

Comment:
If you move the migration of `pin` above the combination check then the check can be much easier.
(plus some braces maybe)
```py
if "version" in attrib
    + "origin" in attrib
    + 0 != len([a for a in attrib if a.startswith('release-')]):

    ...
```